streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -192,29 +213,125 @@ public void processInOrder(final K key, final V value, final long timestamp) {
+                final TimeWindow window = new TimeWindow(previousRecordTimestamp + 1, previousRecordTimestamp + 1 + windows.timeDifferenceMs());
+                final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(initializer.apply(), timestamp);
+                putAndForward(window, valueAndTime, key, value, closeTime, timestamp);
+            }
+
+            if (combinedWindow == null) {
+                final TimeWindow window = new TimeWindow(0, windows.timeDifferenceMs());
+                final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(initializer.apply(), timestamp);
                 putAndForward(window, valueAndTime, key, value, closeTime, timestamp);
+
+            } else {
+                //update the combined window with the new aggregate
+                putAndForward(combinedWindow.key.window(), combinedWindow.value, key, value, closeTime, timestamp);
             }
+
         }
 
-        private boolean rightWindowIsNotEmpty(final ValueAndTimestamp<Agg> rightWinAgg, final long timestamp) {
-            return rightWinAgg != null && rightWinAgg.timestamp() > timestamp;
+        private void createCurrentRecordRightWindow(final long timestamp,
+                                                    final ValueAndTimestamp<Agg> rightWinAgg,
+                                                    final K key) {
+            final TimeWindow window = new TimeWindow(timestamp + 1, timestamp + 1 + windows.timeDifferenceMs());
+            windowStore.put(
+                key,
+                rightWinAgg,
+                window.start());
+            tupleForwarder.maybeForward(
+                new Windowed<>(key, window),
+                rightWinAgg.value(),
+                null,
+                rightWinAgg.timestamp());
+        }
+
+        private boolean leftWindowNotEmpty(final long previousTimestamp, final long currentTimestamp) {
+            return currentTimestamp - windows.timeDifferenceMs() <= previousTimestamp;
+        }
+
+        // previous record's right window does not already exist and current record falls within previous record's right window
+        private boolean rightWindowNecessaryAndPossible(final Set<Long> windowStartTimes,
+                                                        final long previousRightWindowStart,
+                                                        final long currentRecordTimestamp) {
+            return !windowStartTimes.contains(previousRightWindowStart) && previousRightWindowStart + windows.timeDifferenceMs() >= currentRecordTimestamp;
         }
 
-        private boolean isLeftWindow(final KeyValue<Windowed<K>, ValueAndTimestamp<Agg>> window) {
-            return window.key.window().end() == window.value.timestamp();
+        private boolean rightWindowIsNotEmpty(final ValueAndTimestamp<Agg> rightWinAgg, final long timestamp) {
+            return rightWinAgg != null && rightWinAgg.timestamp() > timestamp;
```

Comment:
In our scan for `rightWinAgg` , there's a possibility that we find a `rightWinAgg` that doesn't actually fall within the current record's right window. For example, if we have a record come in at `18`, previous records at `16` and `10`, we would find the window [11,21] (`10`'s right window) when scanning for our `rightWinAgg` in `processInOrder`. [11,21] contains only `16`, which is smaller than `18`, our current record, indicating that the rightWinAgg is incorrect and the right window is in fact empty.

Reply:
Thanks. As above, maybe add a short comment? Can it be that we overload the term "right window"? For this case, `rightWinAgg` seems to be the aggregate of last window "left of the current record" (and it might contribute to the agg of the right window of the current record). However, it "is" not the "right window of the current record" itself?

Reply:
Just added a comment similar to above.
And yeah, we definitely overload the term "right window". The `rightWinAgg` holds what will be in the current record's right window, if anything, but is not _really_ the right window. A name change could be possible, something like `previousWindow` but I wonder if that would make using that aggregate to create the new record's right window more confusing.

Reply:
I thought it was called `rightWinAgg` because it's the aggregate that goes in the current record's right window. Of course we had to find this aggregate from some other existing window, eg from the "last window left of the current record".

Reply:
I think the catch is that it's only _sometimes_ the aggregate that goes in the current record's right window. Sometimes we don't use the value in `rightWingAgg` because the right window is empty/doesn't exist, but if there _is_ a right window aggregate, it will be in `rightWinAgg`